ipnsfs/dir.go

```diff
@@ -0,0 +1,347 @@
+package ipnsfs
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"sync"
+
+	dag "github.com/jbenet/go-ipfs/merkledag"
+	ft "github.com/jbenet/go-ipfs/unixfs"
+	ufspb "github.com/jbenet/go-ipfs/unixfs/pb"
+)
+
+var ErrNotYetImplemented = errors.New("not yet implemented")
+var ErrInvalidChild = errors.New("invalid child node")
+
+type Directory struct {
+	fs        *Filesystem
+	parent    childCloser
+	childDirs map[string]*Directory
+	files     map[string]*file
+
+	node *dag.Node
+	name string
+	lock sync.Mutex
+
+	ref     int
+	refLock sync.Mutex
+}
+
+func NewDirectory(name string, node *dag.Node, parent childCloser, fs *Filesystem) *Directory {
+	return &Directory{
+		fs:        fs,
+		name:      name,
+		node:      node,
+		parent:    parent,
+		childDirs: make(map[string]*Directory),
+		files:     make(map[string]*file),
+	}
+}
+
+func (d *Directory) Open(tpath []string, mode int) (File, error) {
+	if len(tpath) == 0 {
+		return nil, ErrIsDirectory
+	}
+	if len(tpath) == 1 {
+		fi, err := d.childFile(tpath[0])
+		if err == nil {
+			return fi.withMode(mode), nil
+		}
+
+		if mode|os.O_CREATE != 0 {
+			fnode := new(dag.Node)
+			fnode.Data = ft.FilePBData(nil, 0)
+			nfi, err := NewFile(tpath[0], fnode, d, d.fs)
+			if err != nil {
+				return nil, err
+			}
+			d.files[tpath[0]] = nfi
+			return nfi.withMode(mode), nil
+		}
+
+		return nil, ErrNoSuch
+	}
+
+	dir, err := d.childDir(tpath[0])
+	if err != nil {
+		return nil, err
+	}
+	return dir.Open(tpath[1:], mode)
+}
+
+// consider combining into a single method...
+type childCloser interface {
+	closeChild(string, *dag.Node) error
+}
+
+func (d *Directory) closeChild(name string, nd *dag.Node) error {
+	_, err := d.fs.dserv.Add(nd)
+	if err != nil {
+		return err
+	}
+
+	d.lock.Lock()
+	err = d.node.RemoveNodeLink(name)
+	if err != nil && err != dag.ErrNotFound {
+		d.lock.Unlock()
+		return err
+	}
+
+	err = d.node.AddNodeLinkClean(name, nd)
+	if err != nil {
+		d.lock.Unlock()
+		return err
+	}
+	d.lock.Unlock()
```

Comment:
@cryptix i think the lock must be unlocked before closeChild. 

one pattern i like for doing the defer in cases like these is:

```go
func() {
  lk.Lock()
  defer lk.Unlock()
  
  .....
}()

closeChild(...)
```